### ISSUE TITLE
Ensure data loaded into glue is the "right" data to view in Imviz

### ISSUE THREAD

**Issue by eteq** [spacetelescope/jdaviz#600]
To achieve the goal of a being able to use the updated plan for the data drop-down to view extensions, etc, the following is desired as the behavior for what happens when a user loads a multi-extension FITS file:

* Each extension should get loaded into glue as  a data collection where the collection has the file name, and the data objects themselves are individual extensions.
* `imviz.load_data` should default to "load everything", but maybe should have a way to say "just load a particular named extension"
* The Imviz helper should follow the idiom that the rest of jdaviz follows, that `imviz.get_images()` or similar yields a list of `nddata` objects using the glue-astronomy translators.  This should basically be a thin wrapper around `imviz.app.get_data_from_viewer('image-viewer-1')` or similar

An asdf file should follow basically the same convention.

Asdf-in-fits... Uhm.  I honestly am not sure, but I think probably the right answer is to treat it as an asdf file if it *only* includes asdf data, but if it's asdf + fits mixed together, maybe load the asdf extensions with the asdf loader and the fits extensions as first extensions?  I.e. sort of "unroll them". @pllim might have insight though.

And if the user has a nddata or even just a raw 2d array in memory (although the "raw array" case of course wouldn't support WCS), the following should happen:
* `imviz.load_data(nddorarray)` should create a single data collection with a single item in it.  I suppose `imviz.load_data(nddorarray, 'some sort of name')` should be possible too.
* `imviz.load_data([nddorarray1, nddorarray2, ...])` should produce a single collection with multiple data objects 

https://jira.stsci.edu/browse/JDAT-1486

**Comment by eteq** [comment]
(Note if the "data collection" terminology is a bit mystifying, check out http://docs.glueviz.org/en/stable/python_guide/data_tutorial.html, which explains glue's model a bit more)

**Comment by pllim** [comment]
1. How does a data collection play into multiple viewers? What if I have 15 files and I want all the SCI extensions in one viewer, and all the DQ extensions in another viewer?
2. "Load everything" is dangerous for big data. "Just load some" needs clear rules. And when you have rules like that, how do you make exception to the rules? This request needs to be better scoped.
3. `nddata` doesn't play nice with GWCS, see spacetelescope/gwcs#362 . That means, right now, you have to choose between ASDF support and NDData support, you cannot have both.

> ASDF-in-FITS

As far as we're concerned, it is only JWST data, and that is parsed using `jwst.datamodels`. If I see `ASDF` extension in a FITS file, I farm it over to `jwst`. That's about it.

> Programmatically load raw arrays

I think that is already possible with existing API, except maybe for `imviz.load_data([nddorarray1, nddorarray2, ...])`. The latter might solve itself naturally as we implement the other stuff.

**Comment by pllim** [comment]
p.s. ASDF-in-FITS support shouldn't be an afterthought. It is probably the killer feature that would get people to switch from DS9 to Jdaviz because they have no other options (well, they can use Ginga but we already went over that...). I would like some use cases from POs on how you want to load multi-extension ASDF-in-FITS.

**Comment by pllim** [comment]
A few cases of FITS extensions below (non-exhaustive).

# Single-extension

```
Filename: s12_ch1_v0.30_sci.fits
No.    Name      Ver    Type      Cards   Dimensions   Format
  0  PRIMARY       1 PrimaryHDU      35   (2220, 2400)   float32 
```

# Multi-extension (HST/ACS, not corrected for distortion)

```
Filename: jbt7a3k7q_flc.fits
No.    Name      Ver    Type      Cards   Dimensions   Format
  0  PRIMARY       1 PrimaryHDU     281   ()      
  1  SCI           1 ImageHDU       249   (4096, 2048)   float32   
  2  ERR           1 ImageHDU        56   (4096, 2048)   float32   
  3  DQ            1 ImageHDU        48   (4096, 2048)   int16   
  4  SCI           2 ImageHDU       245   (4096, 2048)   float32   
  5  ERR           2 ImageHDU        56   (4096, 2048)   float32   
  6  DQ            2 ImageHDU        48   (4096, 2048)   int16   
  7  HDRLET        1 NonstandardExtHDU     22   (60480,)      
  8  HDRLET        2 NonstandardExtHDU     22   (60480,)      
  9  HDRLET        3 NonstandardExtHDU     26   (112320,)      
 10  HDRLET        4 NonstandardExtHDU     26   (112320,)      
 11  HDRLET        5 NonstandardExtHDU     26   (112320,)      
 12  WCSCORR       1 BinTableHDU     59   14R x 24C   [40A, I, A, 24A, 24A, 24A, 24A, D, D, D, D, D, D, D, D, 24A, 24A, D, D, D, D, J, 40A, 128A]   
 13  WCSDVARR      1 ImageHDU        15   (64, 32)   float32   
 14  WCSDVARR      2 ImageHDU        15   (64, 32)   float32   
 15  D2IMARR       1 ImageHDU        15   (64, 32)   float32   
 16  D2IMARR       2 ImageHDU        15   (64, 32)   float32   
 17  WCSDVARR      3 ImageHDU        15   (64, 32)   float32   
 18  WCSDVARR      4 ImageHDU        15   (64, 32)   float32   
 19  D2IMARR       3 ImageHDU        15   (64, 32)   float32   
 20  D2IMARR       4 ImageHDU        15   (64, 32)   float32   
 21  HDRLET        6 NonstandardExtHDU     26   (112320,)      
 22  HDRLET        7 NonstandardExtHDU     26   (112320,)      
 23  HDRLET        8 NonstandardExtHDU     26   (112320,) 
```

# Multi-extension (HST/ACS, drizzled to correct for distortion)

```
Filename: jbt7a3020_drz.fits
No.    Name      Ver    Type      Cards   Dimensions   Format
  0  PRIMARY       1 PrimaryHDU     758   ()      
  1  SCI           1 ImageHDU        89   (4241, 4253)   float32   
  2  WHT           1 ImageHDU        47   (4241, 4253)   float32   
  3  CTX           1 ImageHDU        40   (4241, 4253)   int32   
  4  HDRTAB        1 BinTableHDU    633   4R x 312C   [9A, 3A, K, D, D, D, D, D, D, D, D, D, D, D, D, D, K, 7A, 9A, 7A, 18A, 4A, D, D, D, D, 3A, D, D, D, D, D, D, D, D, D, D, D, D, K, 8A, 23A, D, D, D, D, K, K, K, 8A, K, 23A, 9A, 21A, K, 4A, K, D, K, K, K, K, 23A, D, D, D, D, K, K, 3A, 3A, 4A, 4A, L, D, D, D, 3A, 1A, K, D, D, D, D, D, 4A, 4A, 12A, 12A, 23A, 8A, 23A, 10A, 10A, D, D, 3A, 3A, 23A, 4A, 8A, 7A, 23A, D, K, D, 6A, 9A, 8A, D, D, L, 4A, 18A, 3A, K, 7A, 5A, 3A, D, 13A, 8A, 4A, 3A, L, K, L, K, L, K, K, D, D, D, D, D, D, 3A, 1A, D, 23A, D, D, D, 3A, 23A, L, 1A, 3A, 6A, D, 3A, 6A, K, D, D, D, D, D, D, D, D, D, D, 23A, D, D, D, D, 3A, D, D, D, 1A, K, K, K, K, K, K, 23A, K, 5A, 7A, D, D, D, D, D, D, D, D, D, D, D, D, D, D, D, D, D, 12A, D, 24A, 23A, D, 1A, 1A, D, K, D, D, 1A, 1A, D, 4A, K, D, K, 7A, D, D, D, D, D, 23A, 23A, D, 8A, D, 29A, D, 3A, D, L, D, D, 3A, 4A, 7A, 1A, D, 7A, K, 1A, 1A, 1A, 1A, D, D, D, D, D, D, 4A, D, 4A, D, 4A, K, 4A, 3A, 1A, L, K, K, 1A, D, D, D, D, K, 3A, L, L, 6A, L, D, D, 13A, D, D, 3A, 8A, 1A, D, K, D, L, 29A, L, 5A, 4A, K, D, D, D, D, D, D, D, D, D, D, D, D, D, D, 12A, 12A, 3A, 3A]
```

# ASDF-in-FITS (JWST/NIRCam, Level 2)

They might add more extensions to it.

```
Filename: jw01072001001_01101_00001_nrcb1_cal.fits
No.    Name      Ver    Type      Cards   Dimensions   Format
  0  PRIMARY       1 PrimaryHDU     235   ()      
  1  SCI           1 ImageHDU        55   (2048, 2048)   float32   
  2  ERR           1 ImageHDU        10   (2048, 2048)   float32   
  3  DQ            1 ImageHDU        11   (2048, 2048)   int32 (rescales to uint32)   
  4  AREA          1 ImageHDU         9   (2048, 2048)   float32   
  5  VAR_POISSON    1 ImageHDU         9   (2048, 2048)   float32   
  6  VAR_RNOISE    1 ImageHDU         9   (2048, 2048)   float32   
  7  VAR_FLAT      1 ImageHDU         9   (2048, 2048)   float32   
  8  ASDF          1 BinTableHDU     11   1R x 1C   [13359B]   
```

**Comment by pllim** [comment]
And I forgot who gave me this file, this is technically a 2D image but it has 4 dimensions. Are we supporting this?

```
Filename: NGC_628_NA_MOM2_THINGS.FITS
No.    Name      Ver    Type      Cards   Dimensions   Format
  0  PRIMARY       1 PrimaryHDU     632   (1024, 1024, 1, 1)   float32 
```

**Comment by eteq** [comment]
> ASDF-in-FITS support shouldn't be an afterthought

Totally agree! It was an "afterthought" in my post above only because it's *more complicated* so I was starting with the simipler one. 

With that in mind - @pllim, for that ASDF-in-FITS example, do you think you can also drop what `jwst.datamodels` produces for that specific file?

**Comment by pllim** [comment]
`jwst.datamodels` returns some sort of, uh, data model that is basically a complicated dictionary. You need to know what key/attribute you want.

**Comment by eteq** [comment]
An alternative that @astrofrog suggested is to have the same general structure, but instead of using labels, the "groups" I described above can be individual glue data objects with the "data objects" in my proposal above instead being *attributes*.  I thought about this but decided against it for the following reasons:

1. they then all have to be the same shape, which as @pllim's "not corrected" case shows, is not always true (there are work-arounds, but they felt awkward as I wrote them out)
2. In FITS files, at least, every HDU can have its own distinct WCS.  This is sometimes true and sometimes not "in the wild", but if I understand right, separate WCSs require separate Data objects.

But if @astrofrog has insight on why those aren't true and/or we might be creating more difficulties this way, I'm all ears!

**Comment by pllim** [comment]
Re: ASDF-in-FITS data file -- Funny you asked `coz I used the file you told me to use. 😸 

https://github.com/spacetelescope/jdaviz/blob/a5c003f4a29e0e48c6e0cee31bad62852ccfbbdc/jdaviz/configs/imviz/tests/test_parser.py#L55

**Comment by astrofrog** [comment]
Having one Data object per HDU is fine and indeed as you said @eteq each HDU might have a different WCS even if they have the same shape. So splitting each HDU into its own Data object makes sense on the glue side.

**Comment by pllim** [comment]
> nddata object with more of the nddata attribute

@eteq , your example won't load because `ndim != 2`. Did you mean to give a 3D example?

**Comment by pllim** [comment]
> nddata object with more of the nddata attribute

Also, if it comes with `.unit`, do I somehow have to project that unit to its `uncertainty` also, or I don't have to?

**Comment by pllim** [comment]
> `imviz.load_data([nddorarray1, nddorarray2, ...])` should produce a single collection with multiple data objects

@eteq , I don't understand this requirement. I don't know how to label them in a way that they are still in the "single collection". They would be labeled like `array1[data]`, `array1[mask]`, `array2[data]`, `array2[mask]` to ensure unique labels anyway, so at that point, why not just use the API to load a single `NDData` object in a loop in the notebook?

**Comment by eteq** [comment]
Oh, good point @pllim - I think you're probably right now that we're not using the collections. But I do think both of the following should be possible:
```
for ndd in nddatas:
    imviz.load_data(ndd)
```
```
for ndd in nddatas:
    imviz.load_data(ndd, name='target ' + ndd.meta['some_special_keyword'])
```
and in the first case you'd get out `array1[data]`, `array1[mask]`, `array2[data]`, `array2[mask]` since it would use whatever the default naming scheme is when the user doesn't give a specific name.

**Comment by eteq** [comment]
> Do we need ".fits"? Currently I have the parser drop ".fits" from label name because I think it is redundant. But I can add it back if you think otherwise.

Hmm.  I think it doesn't really matter since they're almost always unique, but it's not totally unimaginable that someone is, say, comparing an image they downloaded in `asdf` to an image they downloaded in `fits`, so I think I'm mildly in favor of keeping the `.fits`.

> I think it is simpler to just have EXTVER no matter what.

Yes, great point, I hadn't thought of the multiple-SCI's or whatever case, but you're absolutely right.

> What is the value in loading all of these? Takes up memory and usually people only care about SCI extension. What is the use case where you need all the ImageHDU?

I don't think it's true they usually only care about SCI. In the drizzle example you're showing, WHT is critical information for interpreting some things, and so is DQ or ERR in the undrizzled case. I think in general if you load a file the user expects to see what's in that file.  Moreover, there are basically no real standards for these names - `SCI` is an STScI convention, but most other fits files have some other name that has the same meaning as `SCI`, but doesn't mean the same thing.

That said: you raise a good point that we should have an *option* to "load only certain extensions" to save memory especially when there are very large files. For now I think we can say that's available by loading in a notebook and then doing `imviz.load_data(fitsfile['SCI'])` or similar.  But perhaps we should make an issue to design and implement a UI to select extensions?

**Comment by pllim** [comment]
> comparing an image they downloaded in asdf to an image they downloaded in fits

I don't think that is possible. With ASDF-in-FITS, both are in the same file. Really, the only file you can pass into Imviz is `.fits` or some variation of it. I would argue that `somefile.fits` and `somefile.fits.gz` are the same image anyway, if not, there is a deeper problem.

tl;dr -- We should not keep ".fits" in label name

**Comment by PatrickOgle** [comment]
@eteq Is this discussion concluded, or actionable in the near future?  Can we get a synopsis of the way forward?

**Comment by pllim** [comment]
I cannot find the JIRA counterpart to this issue, so I opened https://jira.stsci.edu/browse/JDAT-1486

**Comment by pllim** [comment]
@eteq , reading back the discussions above, I think we have agreed that supporting `imviz.load_data([nddata1, nddata2, ...])` use case is no longer needed. And you didn't disagree on not including extension in label name, so I'd consider that resolved for now.

However, I still strongly disagree that Imviz should load everything by default. Perhaps PvP Round 2 is in order.

**Comment by eteq** [comment]
@pllim, yep, I agree on the first two items (maybe with the caveat on the second one that if e.g. there's a `file.fits` that's been loaded and the user tries to load `file.asdf`, maybe they get a `file.asdf` instead of clobbering `file`, since there are real use cases where the files are different but the names pre-extension are the same.  But that's not critical, since one can always re-name a file :shrug:).

On the "load everything by default"... I also still think it should, so perhaps you are right :wink:

**Comment by pllim** [comment]
Re: `file.fits` vs `file.asdf` -- As far as JWST is concerned, it is not ever going to happen. There is no `.asdf` file being generated.

**Comment by pllim** [comment]
But if you feel strongly about including extension in label, I can add it back. It is not hard to do.

**Comment by eteq** [comment]
Alright, after some out-of-band discussion, I think @pllim and I came to an agreement on the path forward here on the topic of "load everything by default":

My prime concern is that some use cases are for "data exploration" where you want to load all the extensions, and that users will be confused if they load a file with many extensions but only get one of them. @pllim's concern is that loading all the data in a file by default is memory-wasteful and in some cases might even make some files un-load-able as a result. But we agreed that the ideal solution is that explicit is better than implicit. This means the following should be possible from the command line: ``jdaviz --layout imviz filename.fits[*]`` or ``jdaviz --layout imviz filename.fits[SCI]``, which load *all* or *just 1* extension, respectively. (the latter I believe is already implemented by @pllim?)

However, that still leaves the question of what happens if you just do `filename.fits`. Currently it loads only the first valid image extension. While I wouldn't say we "agree", @pllim convinced me I can settle with that behavior *if* we also make sure that in the case where there's other extensions, some sort of prominent message appears in the command line, as a "snackbar" message, or better yet, both, which tells the user that there's some extensions that weren't loaded and says that they can use the `[*]` syntax to load them.

The above is only relevant for the "app interface" perspective - in the notebook it's not that bad to require the user to just for-loop over hdus. But given the above, probably the `imviz.load_data('filename.fits[*]') should be implemented just so that the command-line and the notebook don't have divergent implementations.

We also discussed  "Re: file.fits vs file.asdf" - I agree that's a pretty uncommon case, but there are some real cases, like for example `file.fits` vs `file.jpg`. Or @pllim's point mentioned in https://github.com/spacetelescope/jdaviz/issues/600#issuecomment-856072036 . For both of those, I think a reasonable first iteration is "add something to the end" - i.e. the first one would load as `file`, and the second time would be `file_2` or something.  

Feel free to speak-up ofI mis-remembered any of that, @pllim !

**Comment by pllim** [comment]
> `file_2_2`

I think we only want to try to be smart and append a number if we know user did not explicitly provide a `data_label` to overwrite that behavior. I need to think about this because the way I implemented it, it is tough to differentiate.

I'll post more pointers on 2021-06-10.

**Comment by pllim** [comment]
Notes for @ibusko below. Features 1 and 2 can be separate PRs or in the same one, it is your call. But Feature 3 definitely should be tackled separately after Features 1 and 2 are done.

#  Feature 1: filename.fits[*] to load everything

This can be done by adding special handling of `ext='*'` case here for both plain FITS and ASDF-in-FITS.

https://github.com/spacetelescope/jdaviz/blob/e0efb4f2f7703b869437bd03feb9701b85a93a0b/jdaviz/configs/imviz/plugins/parsers.py#L70

Care need to be given to only seek out all valid image extensions in the `HDUList` and not crashing when the list has other stuff (if you can load a FLT/FLC with those headerlets and a sample Level 2 NIRCam image, should be good enough). New tests need to be added as well along with this work in `jdaviz/configs/imviz/tests/test_parser.py`.

# Feature 1.5: filename.fits[SCI] to only load SCI

Already possible. No new work needed.

# Feature 2: filename.fits to emit message when there are more

Currently, it already loads the first valid image it can find, so no new work needed there. But you do need to add logic to search other other valid image extensions it is ignoring and emit a message. This will make things less performant because it can no longer `break` on first image. `found` can probably be re-purposed for this new logic by changing it from `bool` to `int`. The message hopefully can be fired via some API in the `app` that is getting passed in (unproven).

Not sure how to add a test for this, especially if you emits a "snackbar" message. You probably know better than I do.

https://github.com/spacetelescope/jdaviz/blob/e0efb4f2f7703b869437bd03feb9701b85a93a0b/jdaviz/configs/imviz/plugins/parsers.py#L82

# Feature 3: Incrementing sequence number in data_label for duplicates

This is giving me a headache, because you probably only want to do this if the `data_label` is being generated by the "app" and not be directly given by the user in a notebook cell like `imviz.load_data(..., data_label='mylabel')`. Until I know how to differentiate between the two usage, I cannot give a solid recommendation.

For instance, is it sufficient to say that if `parser_reference is None`, then it is always called by the user and not the app? I have no idea; do you know?

https://github.com/spacetelescope/jdaviz/blob/e0efb4f2f7703b869437bd03feb9701b85a93a0b/jdaviz/configs/imviz/helper.py#L14

Need new tests in `jdaviz/configs/imviz/tests/test_parser.py` too.

**Comment by ibusko** [comment]
@pllim @eteq For feature #2, there is apparently a bug in the way snackbar messages are handled by the main app. Snackbar messages emitted one after the other in a row are not displayed, except the last one. That is, the last snackbar overplots the others that were eventually emitted and displayed before it. That makes all snackbar warning and information messages emitted by the load parsers to disappear from view when the main app emits the "Data successfully loaded" message. Thus, snackbars are an ineffective way to convey user warnings at the time of file loading. Any suggestions on how we should fix this?

**Comment by pllim** [comment]
Are we allowed to change the behavior of `app.load_data` to add a keyword to overwrite this? That way, we can emit "Data successfully loaded (but wait, there are more)" or something all at one go.

https://github.com/spacetelescope/jdaviz/blob/2dfb48775db9b957b6dc0c1b52fe0d7b6a6786bd/jdaviz/app.py#L264

**Comment by pllim** [comment]
FWIW, personally I don't think that "Data successfully loaded" message is useful at all. I mean, it is pretty obvious to my eyes already when the data is loaded. I don't need a message to click through to see it.

**Comment by ibusko** [comment]
What I can do is temporarily disable the "Data successfully loaded" message. And then let the group and POs and whoever, comment on this and reach a conclusion. We can have a PR where people can test it. Bear in mind that that message is application-wide, thus it will affect mosviz, cubeviz, and specviz. Maybe there are workflows where people load data but don't want to see the result visually? (I don't know if that's possible though)

**Comment by ibusko** [comment]
I take it back (sort of). Besides that top-level message, there are others. Say, when a data object is added to a data colection, it emits a snackbar with something like "Data <file-name> successfully added". Clearly, the solution is more complicated.

**Comment by pllim** [comment]
`lock` might work. Or if the app allows stacking, like you see multiple "snackbar" messages stacked in a column like this:

`Message 1 [close]`
`Message 2 [close]`
`Message 3 [close]`

**Comment by ibusko** [comment]
yeah, exactly that. But that's a matter for another ticket. For now, I am going to implement the logic for our particular issue with the fits extensions and test it. It will be acting as a no-op since the snackbar won't be visible unless I comment out the code that creates the other snackbars.

**Comment by pllim** [comment]
Or we can tell Erik that his fantasy isn't possible and people should just use `fitsinfo`. 😸

**Comment by pllim** [comment]
Update: I have marked #680 to close this issue out. I think everything requested would have been implemented when #680 is merged _except_ `imviz.get_images()` (more below).

@eteq , I cannot find any other viz implementation of `.get_images()` anywhere in `jdaviz`. Even if it existed, it is not part of data _loading_ and therefore out of scope. So, I am going to ignore it here. If you feel strongly that this should be implemented as part of MVP, please open a new issue and point me to the actual code in other viz that you mentioned because I cannot find it. Thanks!
